### DISCUSSION TITLE
Why do auth urls point to Supabase servers?

### DISCUSSION THREAD

**Discussion by Nick-Mazuk** [supabase/supabase#1789] [Questions] [answered]
When a user needs to verify their email or reset their password, they get an auth url sent to their email. This url points to a Supabase server. Why?

The current process as I understand it is:

1. User clicks on link in email with a token that identifies the user
2. User goes to Supabase site, where the token is verified
3. User is redirected to our site url with a new token to identify the user
4. Now the user is either verified or their password can be reset

To me, this seems redundant, not an ideal user experience, and prone to error. In my opinion, the user should be directed immediately to our site, then we deal with that token. It also [makes redirection](https://github.com/supabase/supabase/discussions/600) trivial—just change the domain in the link. Firebase works this way, so it should be possible.

- Multiple necessary redirects means slower load times for user
- There are multiple, redundant tokens
- Users may be suspicious of "password reset" emails that go to a site they've never heard of
  - This is a red flag for scam emails
- If the redirection on the Supabase site isn't working for some reason, users won't be redirected to the right place

The only benefit I see is that it eliminates the need for developers setup email verification. But that's really not hard—only 1 extra line of code for the hypothetical `supaase.auth.checkEmailVerificationToken(token)`.

Now I may be misunderstanding why this is happening in the first place. So my main question is, why does it work this way?

**Comment by awalias** [comment, marked as answer]
we could redirect directly to the developer's app, but then the developer has to write code to take the refresh token, send it back to gotrue(the supabase auth server) for verification, at which point an access_token is generated, which is then sent back to the developer's app for them to handle. This would make the whole flow slower, and require that the developer write a custom handler, which we view as unnecessary (and often error-prone) work

just to clarify how something like a signup confirmation works in [gotrue](https://github.com/supabase/gotrue):

1. users signs up
2. we create a confirmation token (valid for 24 hours) and embed it within a link in the email: `ref.supabase.co/auth/v1/verify?token=xyz`
3. the user clicks the link which sends us the confirmation token via their browser
4. we verify the confirmation token and redirect them to the developers site and append an access_token in the URL: `your-site.com#access_token=qwertyuiop`
5. user is now "logged in"

important to note that we could embded the access_token directly in the email, but these are only valid for 1 hour, and it would also be hard for us to "confirm" the email with 100% certainty since JWTs can technically be minted from anywhere using the jwt_secret

hope this helps clarify the flow!

**Reply by awalias** [reply]
oh btw we're also working on returning tokens directly from the backend as an option, and also on webhooks so you will have some flexibility soon around having to embed a supabase url in your transactional emails

https://github.com/supabase/gotrue/issues/65
https://github.com/supabase/gotrue/issues/88

**Reply by vogelino** [reply]
Thx @awalias for clarifying. I understand that simply redirecting to the app has too great implications, but I wonder if it would be possible to achieve the same with a DNS CNAME entry. I suspect some email clients to categorize emails coming from supabase as SPAM when using a custom SMTP server because the sender's email differs from the link's domain. Is my logic making sense?
